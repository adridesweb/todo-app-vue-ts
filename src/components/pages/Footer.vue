<template>
  <footer v-if="!isTodoListEmpty" class="footer" id="footer">
    <TodoCount />
    <Filters />
    <button
      v-if="hideClearCompletedButton"
      @click="deleteCompletedTasks"
      class="clear-completed"
    >
      Clear completed
    </button>
  </footer>
</template>

<script lang="ts">
import { computed, defineComponent } from "vue";
import { useRoute } from "vue-router";
import { useTodosStore } from "@/composables";

import TodoCount from "@/components/TodoCount.vue";
import Filters from "@/components/Filters.vue";

export default defineComponent({
  name: "Footer",
  components: {
    TodoCount,
    Filters,
  },
  setup() {
    const {
      isTodoListEmpty,
      currentUncompletedTasks,
      currentActiveTasks,
      deleteCompletedTasks,
    } = useTodosStore();

    const route = useRoute();

    const hideClearCompletedButton = computed(() => {
      return (
        currentUncompletedTasks.value !== currentActiveTasks.value &&
        route.params.filter !== "active"
      );
    });

    return {
      isTodoListEmpty,
      currentUncompletedTasks,
      currentActiveTasks,
      hideClearCompletedButton,
      deleteCompletedTasks,
    };
  },
});
</script>

<style scoped>
.footer {
  color: white;
  font-family: "Space Mono", monospace;
  padding: 10px 15px;
  height: 20px;
  text-align: center;
  display: flex;
  justify-content: space-between;
  margin-top: 24px;
}

.footer:before {
  content: "";
  position: absolute;
  right: 0;
  bottom: 0;
  left: 0;
  height: 50px;
  overflow: hidden;
  box-shadow: 0 1px 1px rgba(0, 0, 0, 0.2), 0 8px 0 -3px #f6f6f6,
    0 9px 1px -3px rgba(0, 0, 0, 0.2), 0 16px 0 -6px #f6f6f6,
    0 17px 2px -6px rgba(0, 0, 0, 0.2);
}

.clear-completed,
html .clear-completed:active {
  position: relative;
  line-height: 20px;
  text-decoration: none;
  cursor: pointer;
  margin-left: auto;
}

.clear-completed:hover {
  text-decoration: underline;
}

@media (max-width: 530px) {
  .footer {
    height: 50px;
  }
}
</style>
