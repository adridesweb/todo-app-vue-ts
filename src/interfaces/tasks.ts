export interface Task {
  id: string | number;
  title: string;
  completed: boolean;
  editing: boolean;
}
