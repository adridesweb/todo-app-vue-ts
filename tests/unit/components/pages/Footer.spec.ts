import { mount } from "@vue/test-utils";
import { v4 as uuid } from "uuid";
import { Task } from "@/interfaces/tasks";

import Footer from "@/components/pages/Footer.vue";
import createVuexStore from "../../mock-data/mock-store";

describe("tests on pages/Footer.vue", () => {
  const emptyTasks = {
    tasks: [],
  };

  const newBasicTask = <Task>{
    id: uuid(),
    title: "New task",
    completed: false,
    editing: false,
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("renders correctly with one task in store", () => {
    const store = createVuexStore({ tasks: [newBasicTask] });
    const wrapper = mount(Footer, {
      global: {
        plugins: [store],
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  test("does not show footer if no tasks are in store", () => {
    const store = createVuexStore(emptyTasks);
    const wrapper = mount(Footer, {
      global: {
        plugins: [store],
      },
    });

    expect(wrapper.find("footer").exists()).toBe(false);
  });

  test("does show footer if there are tasks in store", () => {
    const store = createVuexStore({ tasks: [newBasicTask] });
    const wrapper = mount(Footer, {
      global: {
        plugins: [store],
      },
    });

    expect(wrapper.find("footer").exists()).toBe(true);
  });
});
