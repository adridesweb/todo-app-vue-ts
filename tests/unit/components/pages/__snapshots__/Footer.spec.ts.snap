// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on pages/Footer.vue renders correctly with one task in store 1`] = `
<footer
  class="footer"
  id="footer"
>
  <div
    class="todo-count"
  >
    <strong>
      1
    </strong>
     item left 
  </div>
  <ul
    class="filters"
  >
    <li>
      <router-link-stub
        to="/"
      />
    </li>
    <li>
      <router-link-stub
        to="active"
      />
    </li>
    <li>
      <router-link-stub
        to="completed"
      />
    </li>
  </ul>
  <!--v-if-->
</footer>
`;
